pkg/kubelet/kubelet.go

```diff
@@ -927,41 +926,17 @@ func (kl *Kubelet) pullImage(img string, ref *api.ObjectReference) error {
 
 // Kill all running containers in a pod (includes the pod infra container).
 func (kl *Kubelet) killPod(pod kubecontainer.Pod) error {
-	// Send the kills in parallel since they may take a long time.
-	errs := make(chan error, len(pod.Containers))
-	wg := sync.WaitGroup{}
-	for _, container := range pod.Containers {
-		wg.Add(1)
-		go func(container *kubecontainer.Container) {
-			defer util.HandleCrash()
-			// Call the networking plugin for teardown.
-			// TODO: Handle this without signaling the pod infra container to
-			// adapt to the generic container runtime.
-			if container.Name == dockertools.PodInfraContainerName {
-				err := kl.networkPlugin.TearDownPod(pod.Namespace, pod.Name, dockertools.DockerID(container.ID))
-				if err != nil {
-					glog.Errorf("Failed tearing down the infra container: %v", err)
-					errs <- err
-				}
-			}
-			err := kl.containerManager.KillContainer(container.ID)
-			if err != nil {
-				glog.Errorf("Failed to delete container: %v; Skipping pod %q", err, pod.ID)
-				errs <- err
-			}
-			wg.Done()
-		}(container)
-	}
-	wg.Wait()
-	close(errs)
-	if len(errs) > 0 {
-		errList := []error{}
-		for err := range errs {
-			errList = append(errList, err)
+	// TODO(vmarmol): Consider handling non-Docker runtimes, the plugins are not friendly to it today.
+	container, err := kl.containerManager.GetPodInfraContainer(pod)
+	if err == nil {
+		// Call the networking plugin for teardown.
+		err = kl.networkPlugin.TearDownPod(pod.Namespace, pod.Name, dockertools.DockerID(container.ID))
+		if err != nil {
+			glog.Errorf("Failed tearing down the network plugin for pod %q: %v", pod.ID, err)
```

Comment:
The difference here is that the error would not be reported upstream.  If we absolutely want to return all errors, we could aggregate the errors again before returning. I'm not sure if it's needed and I am okay with it, so I'll leave it up to you :)